### PR TITLE
feat(anomaly): Sub-Dimension Anomaly Filter

### PR DESCRIPTION
**Pull request by Amatullah**
**Changes Proposed:**
1. Add Subdim Filters on Anomaly Page
2. Remove Sub-dimension sub-page from Anomaly

**Technical Design Document:** [Alerts Phase 2 TDD](https://docs.google.com/document/d/1i1lBCyS1RWtuUgLp2S3ZuhPBFE82TAk_LIgQ6Oo3mb8/edit?usp=sharing)

**Files Changed:**
chaos_genius/chaos_genius/views/anomaly_data_view.py

**Changes:**

1. `kpi_anomaly_detection()`

- Send a dictionary of of sub-dimensions and values for that particular KPI (`{dimension: list(values)}`)
- Return Overall or Sub-dim time series based on the `dimension` and `value` selected by the user

2. `_get_dimensions_values()`

- Generate the dictionary of sub-dims and their values
- Query the db to get the unique `series_type` column values from `anomaly_data_output` table
- Format the returned column into a dictionary.

cc: @ChartistDev @Samyak2

### CODE REVIEW

File: chaos_genius/views/anomaly_data_view.py [hunk 1]
```diff
@@ -425,6 +357,46 @@ def kpi_anomaly_retraining(kpi_id):
         return jsonify({"msg": f"retraining failed for KPI: {kpi_id}, KPI id is None"})
 
 
+def _get_dimensions_values(kpi_id):
+    """Creates a dictionaryy of KPI dimension and their values.
```
Comment (Samyak2) [inline chaos_genius/views/anomaly_data_view.py:361]: Spelling xD

File: chaos_genius/views/anomaly_data_view.py [hunk 2]
```diff
@@ -425,6 +357,46 @@ def kpi_anomaly_retraining(kpi_id):
...
+
+    :param kpi_id: ID of the KPI
+    :type row: integer
```
Comment (Samyak2) [inline chaos_genius/views/anomaly_data_view.py:364]: There's no param named `row` here
```suggestion
    :type kpi_id: integer
```

Comment (Samyak2) [reply chaos_genius/views/anomaly_data_view.py:364]: Also, `integer` isn't a python type. `int` would make it clearer.

File: chaos_genius/views/anomaly_data_view.py [hunk 3]
```diff
@@ -425,6 +357,46 @@ def kpi_anomaly_retraining(kpi_id):
...
+    :return dimensions_vals: dictionary of {dimension:list(vals)}
+    :rtype: dictionary
```
Comment (Samyak2) [inline chaos_genius/views/anomaly_data_view.py:366]: `dict` or `Dict` here

File: chaos_genius/views/anomaly_data_view.py [hunk 4]
```diff
@@ -425,6 +357,46 @@ def kpi_anomaly_retraining(kpi_id):
         return jsonify({"msg": f"retraining failed for KPI: {kpi_id}, KPI id is None"})
 
 
+def _get_dimensions_values(kpi_id):
```
Comment (Samyak2) [inline chaos_genius/views/anomaly_data_view.py:360]: Type annotations are missing.
```suggestion
def _get_dimensions_values(kpi_id: int) -> Dict[str, List[str]]:
```

File: chaos_genius/views/anomaly_data_view.py [hunk 5]
```diff
@@ -425,6 +357,46 @@ def kpi_anomaly_retraining(kpi_id):
...
+    """
+    # Get unique list of subdims and values from DB
+    query = (
+        db.session.query(
+            func.distinct(AnomalyDataOutput.series_type)
+        )
+        .filter(
+            (AnomalyDataOutput.kpi_id == kpi_id)
+            & (AnomalyDataOutput.anomaly_type == "subdim")
+        )
```
Comment (Samyak2) [inline chaos_genius/views/anomaly_data_view.py:376]: What do you think about adding a time bound here? Maybe get only the last 60 or 90 days of data, so that the performance wouldn't degrade too much over time.

File: chaos_genius/views/anomaly_data_view.py [hunk 6]
```diff
@@ -425,6 +357,46 @@ def kpi_anomaly_retraining(kpi_id):
...
+
+    # series_type strings are in the format "`dimension` == "value""
+    # split string into dimension and value to get [`dimension`, "value"]
+    dimension_values_list = list(
+        map(lambda dim_val: dim_val.split(" == "), results["distinct_1"].to_list())
+    )
+    for dim_val in dimension_values_list:
+        # remove extra quotation marks
+        dim_val_formatted = list(map(lambda string: string[1:-1], dim_val))
+        dimension_values_dict[dim_val_formatted[0]].append(dim_val_formatted[1])
```
Comment (Samyak2) [inline chaos_genius/views/anomaly_data_view.py:395]: I would suggesting using unpacking instead of indexes
```suggestion
        dimension, values = map(lambda string: string[1:-1], dim_val)
        dimension_values_dict[dimension].append(values)
```

File: chaos_genius/views/anomaly_data_view.py [hunk 7]
```diff
@@ -425,6 +357,46 @@ def kpi_anomaly_retraining(kpi_id):
...
+        current_app.logger.info("No Subdimension Anomaly Found")
+        return {}
+
+    dimension_values_dict = defaultdict(list)
+
+    # series_type strings are in the format "`dimension` == "value""
+    # split string into dimension and value to get [`dimension`, "value"]
+    dimension_values_list = list(
+        map(lambda dim_val: dim_val.split(" == "), results["distinct_1"].to_list())
+    )
```
Comment (Samyak2) [inline chaos_genius/views/anomaly_data_view.py:391]: We aren't doing anything with the pandas dataframe, just converting one column to a list. Is a dataframe still needed here? You could use `.all()` on the query directly to get the list of values.

File: chaos_genius/views/anomaly_data_view.py [hunk 8]
```diff
@@ -4,6 +4,7 @@
 import io
 import time
 from datetime import date, datetime, timedelta
+from collections import defaultdict
```
Comment (Samyak2) [inline chaos_genius/views/anomaly_data_view.py:7]: Import order was wrong here.

```suggestion
from collections import defaultdict
from datetime import date, datetime, timedelta
```

File: chaos_genius/views/anomaly_data_view.py [hunk 9]
```diff
@@ -77,7 +87,7 @@ def kpi_anomaly_detection(kpi_id):
             # remove from here once updated in frontend
             "base_anomaly_id": kpi_id,
         }
-        data["chart_data"]["title"] = kpi_info["name"]
```
Comment (Samyak2) [inline chaos_genius/views/anomaly_data_view.py:80]: Why was this removed?

File: chaos_genius/views/anomaly_data_view.py [hunk 10]
```diff
@@ -62,12 +63,21 @@ def kpi_anomaly_detection(kpi_id):
...
         period = kpi_info["anomaly_params"]["anomaly_period"]
         hourly = kpi_info["anomaly_params"]["frequency"] == "H"
 
         end_date = get_anomaly_output_end_date(kpi_info)
 
-        anom_data = get_overall_data(kpi_id, end_date, period)
+        dimension = request.args.get("dimension", default=None)
+        value = request.args.get("value", default=None)
+
+        if dimension and value:
```
Comment (Samyak2) [inline chaos_genius/views/anomaly_data_view.py:76]: Do you need to check if `dimension` is `Overall` (and `value` is `None`) too?

File: chaos_genius/views/anomaly_data_view.py (hunk 1 above)
Comment (Amatullah) [reply chaos_genius/views/anomaly_data_view.py]: Fixed.

File: chaos_genius/views/anomaly_data_view.py (hunk 2 above)
Comment (Amatullah) [reply chaos_genius/views/anomaly_data_view.py]: Fixed.

File: chaos_genius/views/anomaly_data_view.py (hunk 3 above)
Comment (Amatullah) [reply chaos_genius/views/anomaly_data_view.py]: Fixed.

File: chaos_genius/views/anomaly_data_view.py (hunk 4 above)
Comment (Amatullah) [reply chaos_genius/views/anomaly_data_view.py]: Fixed.

File: chaos_genius/views/anomaly_data_view.py (hunk 6 above)
Comment (Amatullah) [reply chaos_genius/views/anomaly_data_view.py]: Done.

File: chaos_genius/views/anomaly_data_view.py (hunk 8 above)
Comment (Amatullah) [reply chaos_genius/views/anomaly_data_view.py]: Fixed.

File: chaos_genius/views/anomaly_data_view.py (hunk 5 above)
Comment (Amatullah) [reply chaos_genius/views/anomaly_data_view.py:376]: It's a good idea, but there are some caveats to it:
- We might miss some subdims that appeared before the 90 or 60 days mark. Although that won't matter because we only display the anomaly time period data. 
I'll think about this.

File: chaos_genius/views/anomaly_data_view.py (hunk 9 above)
Comment (Amatullah) [reply chaos_genius/views/anomaly_data_view.py:80]: Because it's redundant. Both the` get_overall_data()` and `get_dq_and_subdim_data()` functions set title value. 
Must have been an oversight initially.

File: chaos_genius/views/anomaly_data_view.py (hunk 10 above)
Comment (Amatullah) [reply chaos_genius/views/anomaly_data_view.py:76]: No. Like we decided, for overall, both dimension and value will be set to None. FE won't return anything.

File: chaos_genius/views/anomaly_data_view.py (hunk 5 above)
Comment (Samyak2) [reply chaos_genius/views/anomaly_data_view.py:376]: Yes. Ideally, it would be good to check only for data after the anomaly time period. If that needs additional data or queries, even harcoding 90 days will let it scale better over time.

File: chaos_genius/views/anomaly_data_view.py (hunk 9 above)
Comment (Samyak2) [reply chaos_genius/views/anomaly_data_view.py:80]: Ah ok

File: chaos_genius/views/anomaly_data_view.py (hunk 10 above)
Comment (Samyak2) [reply chaos_genius/views/anomaly_data_view.py:76]: Got it

File: chaos_genius/views/anomaly_data_view.py (hunk 7 above)
Comment (Amatullah) [reply chaos_genius/views/anomaly_data_view.py]: Done.

File: chaos_genius/views/anomaly_data_view.py (hunk 5 above)
Comment (Amatullah) [reply chaos_genius/views/anomaly_data_view.py]: Done.

File: chaos_genius/views/anomaly_data_view.py [hunk 11]
```diff
@@ -425,6 +357,53 @@ def kpi_anomaly_retraining(kpi_id):
...
+        db.session.query(
+            func.distinct(AnomalyDataOutput.series_type)
+        )
+        .filter(
+            (AnomalyDataOutput.kpi_id == kpi_id)
+            & (AnomalyDataOutput.data_datetime >= start_date)
+            & (AnomalyDataOutput.data_datetime <= end_date)
+            & (AnomalyDataOutput.anomaly_type == "subdim")
+        ).all()
+    )
```
Comment (Samyak2) [inline chaos_genius/views/anomaly_data_view.py:385]: There's a type error on line 372: `Expression of type "str" cannot be assigned to declared type "datetime" `. It's because `end_date` was declared with a specific type (`datetime`) in the function parameters. Looks like the type checker does not allow rebinding an identifier with a different type. The only solution I found was using a different name for the variable (and I changed the name of `start_date_str` to be consistent):

```suggestion
    start_date_str = start_date.strftime("%Y-%m-%d %H:%M:%S")
    end_date_str = end_date.strftime("%Y-%m-%d %H:%M:%S")

    # Get unique list of subdims and values from DB
    results = (
        db.session.query(
            func.distinct(AnomalyDataOutput.series_type)
        )
        .filter(
            (AnomalyDataOutput.kpi_id == kpi_id)
            & (AnomalyDataOutput.data_datetime >= start_date_str)
            & (AnomalyDataOutput.data_datetime <= end_date_str)
            & (AnomalyDataOutput.anomaly_type == "subdim")
        ).all()
    )
```

Comment (Samyak2) [reply chaos_genius/views/anomaly_data_view.py:385]: Wait. Does it really need to be a string? I think SQLalchemy can work with a `datetime` too.

**Review by Samyak2 — COMMENTED** [review]
Looks good. A couple of minor suggestions:

File: chaos_genius/views/anomaly_data_view.py (hunk 11 above)
Comment (Amatullah) [reply chaos_genius/views/anomaly_data_view.py:385]: Not sure if it needs to be a string, but I think there were some issues with the datetime formats and the queries giving incomplete data at boundaries. So I think Kartikay just made it uniform across all functions.

File: chaos_genius/views/anomaly_data_view.py [hunk 12]
```diff
@@ -425,6 +357,53 @@ def kpi_anomaly_retraining(kpi_id):
         return jsonify({"msg": f"retraining failed for KPI: {kpi_id}, KPI id is None"})
 
 
+def _get_dimensions_values(
+    kpi_id: int, end_date: datetime, period=90
+) -> Dict[str, List[str]]:
+    """Creates a dictionary of KPI dimension and their values.
+
+    :param kpi_id: ID of the KPI
+    :type kpi_id: int
```
[the remark this answers is not in the document]
Comment (Amatullah) [reply chaos_genius/views/anomaly_data_view.py:366]: Fixed.

**Review by Samyak2 — APPROVED** [review]
LGTM!

**Review by Samyak2 — APPROVED** [review]
LGTM
